src/Telemetry/Microsoft.DotNet.AzureDevOpsTimelineTests/MockDelayedHttpMessageHandler.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.AzureDevOpsTimeline.Tests
+{
+    public class MockExceptionThrowingHandler : HttpMessageHandler
+    {
+        private readonly Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> _sendAsync;
+        private readonly SemaphoreSlim _mutex;
+        private int _throwAfter;
+        private int _counter;
+
+        public MockExceptionThrowingHandler(Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> sendAsync, int throwAfter)
+        {
+            _sendAsync = sendAsync;
+            _throwAfter = throwAfter;
+            _counter = 0;
+            _mutex = new SemaphoreSlim(1);
+        }
+
+        public static MockExceptionThrowingHandler Create(string message, int throwAfter)
+        {
+            return new MockExceptionThrowingHandler(async (request, cancellationToken) =>
+            {
+                var responseMessage = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content = new StringContent(message)
+                };
+
+                return await Task.FromResult(responseMessage);
+            }, throwAfter);
+        }
+
+        public static MockExceptionThrowingHandler Create(int throwAfter)
+        {
+            return new MockExceptionThrowingHandler(async (request, cancellationToken) =>
+            {
+                var responseMessage = new HttpResponseMessage(HttpStatusCode.OK);
+
+                return await Task.FromResult(responseMessage);
+            }, throwAfter);
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            _mutex.Wait(cancellationToken);
```

Comment:
I think just incrementing isn't enough, both incrementing and comparing have to be done atomically I think

Reply:
The number can't go down, and "_throwAfter" is immutable. "Interlocked.Increment" returns the value that it incremented, so that you can do the comparison.

If you really don't want to do the interlocked thing, a Semaphore of 1 is just a "new object" that you `lock`.

Reply:
I removed the mutex, using `Interlocked.Increment` now, just had to change the comparison to <= since we are using pre increment now